src/library/Cardano/Fee.hs

```diff
@@ -331,26 +347,84 @@ coverRemainingFee (Fee fee) = go [] where
 -- >>> reduceChangeOutputs (DustThreshold 0) (Fee 15) (Coin <$> [1, 2, 4, 8])
 -- []
 --
-reduceChangeOutputs :: DustThreshold -> Fee -> [Coin] -> [Coin]
-reduceChangeOutputs threshold (Fee totalFee) changeOutputs
-    | totalFee >= totalChange =
-        []
-    | otherwise =
-        case positiveChangeOutputs of
-            x : xs -> x :| xs
-                & distributeFee (Fee totalFee)
-                & fmap payFee
-                & coalesceDust threshold
-            [] -> []
-  where
-    payFee :: (Fee, Coin) -> Coin
-    payFee (Fee f, c) =
-        fromMaybe C.zero (c `C.sub` f)
-
-    positiveChangeOutputs :: [Coin]
-    positiveChangeOutputs = filter (> C.zero) changeOutputs
-
-    totalChange = F.fold changeOutputs
+reduceChangeOutputs
+    :: FeeOptions i o
+    -> CoinSelection i o
+    -> Either (FeeAdjustmentError i o) (CoinSelection i o, Fee)
+reduceChangeOutputs opts s = do
+    -- The original requested fee amount
+    let Fee φ_original = estimateFee (feeEstimator opts) s
+    -- The initial amount left for fee (i.e. inputs - outputs)
+    let mδ_original = sumInputs s `C.sub` (sumOutputs s `C.add` sumChange s)
+    case mδ_original of
+        -- selection is now balanced, nothing to do.
+        Just δ_original | φ_original == δ_original -> do
+            pure (s, Fee C.zero)
+
+        -- some fee left to pay, but we've depleted all change outputs
+        Just δ_original | φ_original > δ_original && null (change s) -> do
+            let Just remainder = φ_original `C.sub` δ_original -- cf above guard
+            pure (s, Fee remainder)
+
+        -- some fee left to pay, and we've haven't depleted all change yet
+        Just δ_original | φ_original > δ_original && not (null (change s)) -> do
+            let Just remainder = φ_original `C.sub` δ_original -- cf above guard
+            let chgs' = distributeFee (Fee remainder) (NE.fromList (change s))
+                      & fmap payFee
+                      & coalesceDust (dustThreshold opts)
+            reduceChangeOutputs opts (s { change = chgs' })
+
+        -- the selection is leaving more fee than necessary. This typically
+        -- occurs when, after reducing a an output to pay for existing fees,
+        -- fees turn out to be less expensive than originally. In which, the
+        -- behavior will depend on whether or not the node allows transaction to
+        -- be unbalanced or not.
+        Just δ_original | δ_original > φ_original -> do
+            let Just extraChng = δ_original `C.sub` φ_original -- cf above guard
```

Comment:
I notice there are quite a few partial pattern matches. (I can understand why this has happened, as it's very close to the style used in `cardano-wallet` for the equivalent piece of code.)

To avoid a partial pattern match, we could use the `C.distance` function here:
```suggestion
        Just δ_original | δ_original > φ_original -> do
            let extraChng = δ_original `C.distance` φ_original -- cf above guard
```

However, we could just as well use ordinary pattern matching here:
```hs
Just δ_original ->
    case δ_original `C.sub` φ_original of
        Just extraChange ->
            ...
        Nothing ->
            ...
```
**Advantages**:
- shorter code
- ordinary pattern matching is arguably more resistant to regressions (if a future maintainer removes the guard, the compiler won't notice that the partial pattern match is now unguarded).
- we can remove the comment

Reply:
I disagree here as the code is much easier to follow when all branches are at the same level. I consciously chose not to use nested `case .. of` here because it otherwise create a hierarchy of cases that are hard to follow.

Using guards allows for flattening it which result in a code that is much easier to follow, because each case is clearly identifiable by the structure of the code itself. Actually, the version in cardano-wallet is even more flat by not having: 

```
    let mδ_original = sumInputs s `C.sub` (sumOutputs s `C.add` sumChange s)
    case mδ_original of
```

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Primitive/Fee.hs#L265-L301

It releis on lazyness and the fact that the `extraChng` only is evaluated when `δ_original > φ_original`.  The body of the function reads much better, although re-organizing the branches can have a dramatic effect that isn't obvious at first sight. So in the cardano-coin-selection version, I decided to have this first pattern match as a middle-ground, but still keep it flat afterwards :)

Reply:
>  The body of the function reads much better, although re-organizing the branches can have a dramatic effect that isn't obvious at first sight. So in the cardano-coin-selection version, I decided to have this first pattern match as a middle-ground, but still keep it flat afterwards :)

Fair enough. :+1: 

In that case, I think using `C.distance` here rather than `C.sum` would be a good choice here, to avoid the partial match on `Maybe.`